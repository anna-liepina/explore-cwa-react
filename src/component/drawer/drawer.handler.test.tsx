import React from 'react';
import '@testing-library/jest-dom';
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import DrawerHandler from './drawer.handler';

describe('<DrawerHandler/>', () => {
    beforeAll(() => {
        const portalElement = document.createElement('div');
        portalElement.id = 'portal';
        document.body.appendChild(portalElement);
    });

    const props = {
        button: (props: React.ComponentPropsWithRef<'button'>) => <button {...props} />
    };

    const optionalProps = {
        'data-cy': 'optionalProps.data-cy',
        className: 'optionalProps.className',
    };

    describe('render', () => {
        it('with default/required props', () => {
            const { asFragment } = render(<DrawerHandler {...props} />);

            expect(asFragment()).toMatchSnapshot();
        });

        it('with optional/required props', () => {
            const { asFragment } = render(<DrawerHandler {...props} {...optionalProps} />);

            expect(asFragment()).toMatchSnapshot();
        });
    });

    describe('internal callbacks', () => {
        describe('onToggle', () => {
            it('from a click on [data-cy="--drawer--open"] should change internal state, and RENDER <Drawer /> and it\'s content', async () => {
                render(<DrawerHandler {...props} />);

                expect(document.querySelector('[data-cy="--drawer--close"]')).not.toBeInTheDocument();

                act(() => {
                    fireEvent.click(document.querySelector('[data-cy="--drawer--open"]')!);
                });

                await waitFor(() => {
                    expect(document.querySelector('[data-cy="--drawer--close"]')).toBeInTheDocument();
                })
            });
            
            it('from a click on [data-cy="--drawer--close"] should change internal state, and HIDE <Drawer /> and it\'s content', async () => {
                render(<DrawerHandler {...props} isExpanded />);

                expect(document.querySelector('[data-cy="--drawer--close"]')).toBeInTheDocument();

                act(() => {
                    fireEvent.click(document.querySelector('[data-cy="--drawer--close"]')!);
                });

                await waitFor(() => {
                    expect(document.querySelector('[data-cy="--drawer--close"]')).not.toBeInTheDocument();
                })
            });
        });
    });

    describe('external callbacks', () => {
        describe('onToggle', () => {
            it('should be invoked from a click on a [data-cy="--drawer--open"]', () => {
                const spy = jest.fn();
                const { container } = render(<DrawerHandler {...props} onToggle={spy} />);

                act(() => {
                    fireEvent.click(container.querySelector('[data-cy="--drawer--open"]')!);
                })

                expect(spy).toHaveBeenCalled();
            });

            it('should be invoked from a click on a [data-cy="--drawer--close"]', async () => {
                const spy = jest.fn();
                render(<DrawerHandler {...props} onToggle={spy} isExpanded />);

                expect(document.querySelector('[data-cy="--drawer--close"]')).toBeInTheDocument();

                act(() => {
                    fireEvent.click(document.querySelector('[data-cy="--drawer--close"]')!);
                });

                await waitFor(() => {
                    expect(spy).toHaveBeenCalled();
                })
            });
        });
    });
});
