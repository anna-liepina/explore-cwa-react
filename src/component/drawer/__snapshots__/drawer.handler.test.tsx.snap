// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrawerHandler/> render with default/required props 1`] = `
<DocumentFragment>
  <button
    data-cy="--drawer--open"
  />
</DocumentFragment>
`;

exports[`<DrawerHandler/> render with optional/required props 1`] = `
<DocumentFragment>
  <button
    data-cy="optionalProps.data-cy--drawer--open"
  />
</DocumentFragment>
`;
